src/Components/LineLens.fs

```diff
@@ -0,0 +1,376 @@
+    let inline set fileName value (documents: Documents) = documents.Add(fileName, value)
+
+    let update info (decorations: ResizeArray<DecorationOptions>) version (documents: Documents) =
+        let decoratedLines = decorations |> Seq.map (fun d -> d.range.start.line) |> HashSet
+        let updated =
+            { info with
+                cache =
+                    Some {
+                        version = version
+                        decorations = decorations
+                        decoratedLines = decoratedLines } }
+        documents |> set info.fileName updated
+
+let mutable private config = LineLensConfig.defaultConfig
+
+module LineLensDecorations =
+    let create range text =
+        // What we add after the range
+        let attachment = createEmpty<ThemableDecorationAttachmentRenderOptions>
+        attachment.color <- Some config.color
+        attachment.contentText <- Some text
+
+        // Theme for the range
+        let renderOptions = createEmpty<DecorationInstanceRenderOptions>
+        renderOptions.after <- Some attachment
+
+        let decoration = createEmpty<DecorationOptions>
+        decoration.range <- range
+        decoration.renderOptions <- Some renderOptions
+        decoration
+
+    let decorationType =
+        let opt = createEmpty<DecorationRenderOptions>
+        opt.isWholeLine <- Some true
+        opt
+
+type State = {
+    documents: Documents.Documents
+    decorationType: TextEditorDecorationType
+    disposables: ResizeArray<Disposable>
+}
+
+module DecorationUpdate =
+    let private lineRange (doc: TextDocument) (range: DTO.Range) : CodeRange.CodeRange =
+        let lineNumber = float range.StartLine - 1.
+        let textLine = doc.lineAt lineNumber
+        textLine.range
+
+    let private getSignature (fileName: string) (range: DTO.Range) = promise {
+        let! signaturesResult =
+            LanguageService.signatureData
+                fileName
+                range.StartLine
+                range.StartColumn
+        let signaturesResult = if isNotNull signaturesResult then Some signaturesResult else None
+        return signaturesResult |> Option.map (fun r -> range, CodeLens.formatSignature r.Data)
+    }
+
+    let private signatureToDecoration (doc: TextDocument) (range:DTO.Range, signature:string) =
+        LineLensDecorations.create (lineRange doc range) (config.prefix + signature)
+
+    let private onePerLine (ranges: Range[]) =
+        ranges
+        |> Array.groupBy(fun r -> r.StartLine)
+        |> Array.choose (fun (_, ranges) -> if ranges.Length = 1 then Some (ranges.[0]) else None)
+
+    let private updateDecorationsForEditor (textEditor : TextEditor) state = promise {
+        let fileName = textEditor.document.fileName
+        let version = textEditor.document.version
+        let text = textEditor.document.getText()
+
+        // TODO: Other things parse doc, how to wait for that ?
+        let! _ = LanguageService.parse fileName text version
```

Comment:
You are already waiting until the parse request, with the `let!`
Atm the parse is cached FSAC side, so should be fast, but there isnt yet an event in ionide side triggered after a document is parsed (or when fsac has sent a `parse` response)

Reply:
Ok the cache was my main question :) I didn't want to end up with the same version being parsed twice juste because LineLens are enabled.

Reply:
> Atm the parse is cached FSAC side...
This is true partially... as the response will be cached if the `version` was already parsed. 

I guess I would follow same pattern as Linter and CodeLens is using - define refresh event in module (https://github.com/ionide/ionide-vscode-fsharp/blob/master/src/Components/Linter.fs#L16) which is triggered when file is parsed (https://github.com/ionide/ionide-vscode-fsharp/blob/master/src/Components/Errors.fs#L41)

Reply:
If you will update only when this event is called, you will get automatic handling for file open. You'll also be able to drop all the code you have for controlling when to refresh LineLenses

Reply:
Ah thanks, that's exactly what I needed when I started. I wondered where CodeLens got it's updates.

It'll simplify some things & complexify others as I don't want them to appear on lines the user is typing (Decorations are applied asynchronously and can produce bugs where the user type after them)

At least it'll ensure no double-parse

Reply:
CodeLenses are bit more complex as they're also refreshed internally by VSCode, not only on the refresh I call (which causes some problems). But anyway... I'd go with event triggered after parse is completed, at least it will work same as other modules

Reply:
Oh i'll do the change it's better all around. I'm just not sure that it simplifies the logic a lot but it'll work better for sure.

The bit about code lenses auto updating was why I didn't dig too much into the event, I assumed that it was linked to the refresh by vscode not by somewhere else in the extension.